src/queue.rs

```diff
@@ -548,6 +563,74 @@ impl<M: GuestAddressSpace> Queue<M> {
         snapshot
             .write_obj(self.next_used.0 as u16, used_ring.unchecked_add(2))
             .unwrap();
+
+        Some(self.next_used.0)
+    }
+
+    /// Update avail_event on the used ring with the last index in the avail ring.
+    pub fn update_avail_event(&mut self) {
+        // Safe because we have validated the virt queue.
+        // And all are two-byte naturally aligned accesses, so they never cross the memory region
+        // boundary.
+        let mem = self.mem.memory();
+        let index_addr = self.avail_ring.unchecked_add(2);
+        match mem
+            .get_slice(index_addr, size_of::<u16>())
+            .map(|s| s.get_ref::<u16>(0).unwrap().load())
+        {
+            Ok(index) => {
+                let offset = (4 + self.actual_size() * 8) as u64;
+                let avail_event_addr = self.used_ring.unchecked_add(offset);
+                if let Ok(avail_event_slice) = mem.get_slice(avail_event_addr, size_of::<u16>()) {
+                    // This fence ensures the guest sees the value we've just written.
+                    avail_event_slice
+                        .get_atomic_ref::<AtomicU16>(0)
+                        .unwrap()
+                        .store(index, Ordering::Release);
+                } else {
+                    warn!("Can't update avail_event");
+                }
+            }
+            Err(e) => warn!("Invalid offset, {}", e),
+        }
+    }
+
+    /// Return the value present in the used_event field of the avail ring.
+    fn get_used_event(&self, mem: &M::M) -> Option<Wrapping<u16>> {
+        // Following code is safe because we have validated the queue. And the `used_index` is a
+        // two-byte naturally aligned field, so it won't cross the region boundary and get_slice()
+        // shouldn't fail.
+        let used_event_addr = self
+            .avail_ring
+            .unchecked_add((4 + self.actual_size() * 2) as u64);
+        // This fence ensures we're seeing the latest update from the guest.
+        mem.get_slice(used_event_addr, size_of::<u16>())
+            .map(|s| {
+                Wrapping(
+                    s.get_atomic_ref::<AtomicU16>(0)
+                        .unwrap()
+                        .load(Ordering::SeqCst),
```

Comment:
Never use `SeqCst`, it's really hard to use and can lead to extremely complicated bugs. See [here](https://patchew.org/QEMU/20200407140746.8041-1-pbonzini@redhat.com/20200407140746.8041-6-pbonzini@redhat.com/) for an example.

`SeqCst` fences are okay though.

Reply:
Thanks for the reference, it's very helpful. Seems that `Ordering::Relaxed` is safe for the EVENT_IDX related operations. 
Yeap, the `SeqCst` is really tricky, a long description in C11 stand:(

Reply:
I suggest you use an `enable_notification` method similar to QEMU's.